### PR TITLE
Allen/get minimum1

### PR DESCRIPTION
**Pull request by algtan**
This PR supersedes https://github.com/EdgeApp/edge-swap-server/pull/2. Needed to upgrade `edge-server-tools` to fix this PR, so now there is a dependency on PR https://github.com/EdgeApp/edge-swap-server/pull/5. The `allen/getMinimum1` branch is the result of cherry-picking the commits.

Original description:
Added an endpoint, getMinimum, that returns the minimum amount to do a swap for a currency pair. The endpoint checks for the latest documents for the currency pair per swap provider database.

### CODE REVIEW

**Comment by LPeterson18** [comment]
Looks good.

File: src/utils/errorResponse.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { asNumber, asString } from 'cleaners'
+
+export interface ErrorResponse extends Error {
```
Comment (algtan) [inline src/utils/errorResponse.ts:3]: Use a union instead of extending the interface.

File: src/utils/errorResponse.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  message: string
```
Comment (algtan) [inline src/utils/errorResponse.ts:4]: The Error object has a built in `message` property.

File: src/utils/errorResponse.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+export const makeErrorResponse = (
+  errorType: string,
+  errorCode: number,
+  message: string = defaultErrorMessage
+): ErrorResponse => {
+  return {
+    ...new Error(),
+    message: asString(message),
+    errorType: asString(errorType),
+    errorCode: asNumber(errorCode)
```
Comment (algtan) [inline src/utils/errorResponse.ts:20]: Make one line.

File: src/utils/utils.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+export const checkQueryString = (
+  queryValue: string
+): string | ErrorResponse => {
+  if (typeof queryValue !== 'string') {
+    return ParamError
+  }
+  return queryValue
+}
```
Comment (algtan) [inline src/utils/utils.ts:16]: Make a one-liner using the ternary operator.

File: src/utils/getMinimum.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      } = asCouchMinAmountDocData(currentDoc.value)
+      return [...minAmountDocDataArr, data]
+    } catch {
+      return minAmountDocDataArr
+    }
+  }, [])
+}
+
+export const findMinimum = (
+  docData: Array<ReturnType<typeof asMinAmountInfo>>
```
Comment (algtan) [inline src/utils/getMinimum.ts:36]: Find a more descriptive parameter name to replace `docData`.

File: src/indexApi.ts [hunk 6]
```diff
@@ -2,17 +2,81 @@
...
+// ROUTES FOR OUR API
+// =============================================================================
+const router = express.Router()
+
+router.get('/getMinimum/', async function (req, res, next) {
+  try {
+    const currencyPairResult = checkQueryString(req.query.currencyPair) // Param of 'currencyPair' in URI
+    if (typeof currencyPairResult !== 'string') {
+      throw currencyPairResult // currencyPairResult is equal to ParamError
+    }
```
Comment (algtan) [inline src/indexApi.ts:46]: Place in middleware before the router.

File: src/indexApi.ts [hunk 7]
```diff
@@ -2,17 +2,81 @@
...
+    }) // Variable for current UTC date using date-fns
+
+    const minAmtDocKey = currentUTCDate + ':' + currencyPairResult
+    const minAmtDocs = await fetchSameKeyDocs(minAmtDocKey, minAmtDocScopeArr)
+    const cleanMinAmtData = cleanMinAmountDocs(minAmtDocs)
+    const minAmtResult = findMinimum(cleanMinAmtData)
+
+    if (typeof minAmtResult !== 'string') {
+      throw minAmtResult // minAmtResult is equal to CurrencyPairDataError
+    }
```
Comment (algtan) [inline src/indexApi.ts:58]: Place in a new function outside the middleware. The route will then be a one-liner inside the `try..catch`, which will reduce the number of lines for this middleware. NOTE: Try to keep functionality that uses `req, res, next` inside the middleware, and everything outside of it.

**Comment by LPeterson18** [comment]
Looks good to me.

**Review by thehobbit85 — APPROVED** [review]

**Pull request merged**